cpp/src/experimental/graph.cu

```diff
@@ -0,0 +1,543 @@
+                    d_segment_offsets.size(),
+                    handle.get_stream());
+
+  return h_segment_offsets;
+}
+
+}  // namespace
+
+template <typename vertex_t,
+          typename edge_t,
+          typename weight_t,
+          bool store_transposed,
+          bool multi_gpu>
+graph_t<vertex_t, edge_t, weight_t, store_transposed, multi_gpu, std::enable_if_t<multi_gpu>>::
+  graph_t(raft::handle_t const &handle,
+          std::vector<edgelist_t<vertex_t, edge_t, weight_t>> const &edgelists,
+          partition_t<vertex_t> const &partition,
+          vertex_t number_of_vertices,
+          edge_t number_of_edges,
+          bool is_symmetric,
+          bool is_multigraph,
+          bool is_weighted,
+          bool sorted_by_global_degree_within_vertex_partition,
+          bool do_expensive_check)
+  : detail::graph_base_t<vertex_t, edge_t, weight_t>(
+      handle, number_of_vertices, number_of_edges, is_symmetric, is_multigraph, is_weighted),
+    partition_(partition)
+{
+  // cheap error checks
+
+  auto &comm_p         = this->get_handle_ptr()->get_comms();
+  auto comm_p_size     = comm_p.get_size();
+  auto &comm_p_row     = this->get_handle_ptr()->get_subcomm(comm_p_row_key);
+  auto comm_p_row_rank = comm_p_row.get_rank();
+  auto comm_p_row_size = comm_p_row.get_size();
+  auto &comm_p_col     = this->get_handle_ptr()->get_subcomm(comm_p_col_key);
+  auto comm_p_col_rank = comm_p_col.get_rank();
+  auto comm_p_col_size = comm_p_col.get_size();
+  auto default_stream  = this->get_handle_ptr()->get_stream();
+
+  for (size_t i = 0; i < edgelists.size(); ++i) {
+    CUGRAPH_EXPECTS(
+      (edgelists[i].p_src_vertices != nullptr) && (edgelists[i].p_dst_vertices != nullptr),
+      "Invalid API parameter: edgelists[].p_src_vertices and edgelists[].p_dst_vertices should "
+      "not be nullptr.");
+    CUGRAPH_EXPECTS((is_weighted && (edgelists[i].p_edge_weights != nullptr)) ||
+                      (!is_weighted && (edgelists[i].p_edge_weights == nullptr)),
+                    "Invalid API parameter: edgelists[].p_edge_weights should not be nullptr (if "
+                    "is_weighted is true) or should be nullptr (if is_weighted is false).");
+  }
+
+  CUGRAPH_EXPECTS((partition.hypergraph_partitioned &&
+                   (edgelists.size() == static_cast<size_t>(comm_p_row_size))) ||
+                    (!(partition.hypergraph_partitioned) && (edgelists.size() == 1)),
+                  "Invalid API parameter: errneous edgelists.size().");
+
+  CUGRAPH_EXPECTS(partition.vertex_partition_offsets.size() == static_cast<size_t>(comm_p_size),
+                  "Invalid API parameter: erroneous partition.vertex_partition_offsets.size().");
+
+  // optinoal expensive checks (part 1/3)
+
+  if (do_expensive_check) {
+    edge_t number_of_local_edges_sum{};
+    for (size_t i = 0; i < edgelists.size(); ++i) {
+      auto major_first =
+        partition.hypergraph_partitioned
```

Comment:
Consider making this (and the other ternary operators below) a member function of the partition object.  Might make the code a bit cleaner.

Reply:
OK, now partition_t is a class and hypergraph_partitioned & vertex_partition_offsets are private.